KPI Dashboard: add Allure reporter to Jest environment.

worldomonation opened the pull request:
#### Proposed Changes

This PR implements a Jest-Circus friendly Allure reporter into our custom Jest environment.

**Key changes**:
- inclusion of core Allure reporter code from [ryparker/jest-circus-allure-environment](https://github.com/ryparker/jest-circus-allure-environment) under MIT license.
- adapted the above code into a stand-alone TypeScript module.
- added Allure event commands to the Jest test event handler.

**Details and background**:
As part of the [OKRs for 2022](pciE2j-EI-p2), KitKat wants to create a KPI dashboard to visualize end-to-end test performance over time. After evaluating our options, it was decided to try the Allure framework to semi-automatically generate rich reports.

In the investigation I hit a snag in that Jest-Circus runner, default on Jest 27 and what the E2E framework has been built around, is not compatible with Allure out of the box. The many reporters and environments that others have written do not work for our framework for a [variety of reasons](https://github.com/Automattic/wp-calypso/issues/66190#issuecomment-1230801476). 

Therefore, I identified the best course of action was to adapt existing open-source code at [ryparker/jest-circus-allure-environment](https://github.com/ryparker/jest-circus-allure-environment) to our custom Jest environment. As the original source is distributed under an MIT license, I believe this falls under permissible use.

**Implementation Details**:
We extend the default `JestNodeEnvironment` with our own custom setup code:
- create artifact directory and define names for the resulting artifacts;
- set video and screen capture parameters;
- start up and maintain the browser instance.

We also add custom handling to different events fired by Jest as it executes the test script. For instance, by hooking into the `teardown` event we can instruct the global browser instance to terminate, triggering the artifacts to be written to disk. 

To implement Jest-Circus compatible Allure reporting, an instance of an Allure reporter is instantiated within the custom environment setup. Within the event handler, different methods within the Allure reporter is called to generate data. For instance, when the `test_fn_failure` event is fired the Allure reporter is instructed to mark the existing test as failed as well for report generation purposes.

#### Testing Instructions

Ensure the following build configurations are passing:
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E
  - [x] Code Style
  - [x] Unit Tests

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes https://github.com/Automattic/wp-calypso/issues/66190.

worldomonation committed:
Install allure-js-commons

worldomonation committed:
Add new package @automattic/jest-circus-allure-reporter.

worldomonation committed:
Add new package `@automattic/jest-circus-allure-reporter` to dependencies throughout the tree.

worldomonation committed:
packages/calypso-e2e/src/jest-playwright-config/environment.ts
- hook the Allure reporter into our Jest environment.

worldomonation committed:
.teamcity/_self/lib/customBuildType/E2EBuildType.kt
- build package in CI

test/e2e/.gitignore
- add allure-results to git ignore

test/e2e/package.json
- define new scripts for `test/e2e` to make life easier

worldomonation committed:
.teamcity/_self/lib/customBuildType/E2EBuildType.kt
- experimental: add allure-results to the artifacts list

worldomonation committed:
packages/calypso-e2e/tsconfig.json
- add reference to `@automattic/jest-circus-allure-reporter`.

packages/tsconfig.json
- fix alphabetical ordering

worldomonation committed:
.teamcity/_self/lib/customBuildType/E2EBuildType.kt
- in the collect logs step add referemces tp allure-results

packages/calypso-e2e/src/env-variables.ts
- add new environment variable ALLURE_RESULTS_PATH

packages/calypso-e2e/src/types/env-variables.types.ts
- add ALLURE_RESULTS_PATH to typng

packages/calypso-e2e/src/jest-playwright-config/environment.ts
- use ALLURE_RESULTS_PATH when configuring allure

worldomonation committed:
.teamcity/_self/lib/customBuildType/E2EBuildType.kt
- fix folder name

worldomonation commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -29,24 +34,46 @@ const supportedBrowsers = [ chromium, firefox ];
  * @see {@link https://github.com/facebook/jest/tree/main/packages/jest-circus}
  */
-class JestEnvironmentPlaywright extends JestEnvironmentNode {
+class JestEnvironmentPlaywright extends NodeEnvironment {
```
I'm not certain of the difference between `JestNodeEnvironment` and `NodeEnvironment`, but most examples of similar work I've seen on the web use the latter.

worldomonation commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -183,29 +210,59 @@ class JestEnvironmentPlaywright extends JestEnvironmentNode {
 	 * Handle events emitted by jest-circus.
 	 */
-	async handleTestEvent( event: Circus.Event ) {
+	async handleTestEvent( event: Circus.Event, state: Circus.State ) {
```
Calls Allure methods that correspond to the Jest emitted events for fine grained control over the reporting values.

The list of events are available [here](https://github.com/facebook/jest/blob/main/packages/jest-types/src/Circus.ts).

worldomonation commented on packages/jest-circus-allure-reporter/.eslintrc.js:
```diff
@@ -0,0 +1,26 @@
+			'error',
+			{ argsIgnorePattern: '^_', ignoreRestSiblings: true },
+		],
+		'import/no-nodejs-modules': 'off',
```
This is copied from the `@automattic/calypso-e2e` package but with one additional change at this line, as the built-in package `crypto` and `os` are used.

noahtallen replied:
Interesting -- I thought the "nodeConfig" would allow nodeJS modules. (After all, the whole point of that module is to to share the config for any node projects which need it)

worldomonation commented on packages/jest-circus-allure-reporter/README.md:
```diff
@@ -0,0 +1,152 @@
+# Jest Circus Allure Reporter
```
README is simplified from the [original](https://github.com/ryparker/jest-circus-allure-environment/blob/main/README.md):
- removed quick start guide
- removed advanced configuration section
- removed TMS and Jira specifics

Acknowledgement of the origin of this package is inserted at top.

worldomonation commented on packages/jest-circus-allure-reporter/src/index.ts:
```diff
@@ -0,0 +1,5 @@
+
+export * from './types';
+
+export { AllureRuntime, AllureConfig } from 'allure-js-commons';
```
This is to eliminate the need to have `allure-js-commons` as a direct dependency of `@automattic/calypso-e2e`. 

These exports are used in `environment.ts` during the initialization of Allure reporter.

worldomonation commented on packages/jest-circus-allure-reporter/src/reporter.ts:
```diff
@@ -0,0 +1,519 @@
+import { parseWithComments } from 'jest-docblock';
+import { ContentType } from './types';
+
+/**
+ * Represents an instance of an AllureReporter.
+ */
```
This file is based on the [original file](https://github.com/ryparker/jest-circus-allure-environment/blob/main/src/allure-base-environment.ts) but with important changes to make it a standalone class.

Furthermore, JSDoc were added to all methods.

worldomonation commented on packages/jest-circus-allure-reporter/src/step-wrapper.ts:
```diff
@@ -0,0 +1,57 @@
+/* eslint-disable require-jsdoc */
```
In the interest of getting this code reviewed sooner I have disabled the `require-jsdoc` rule for this file.

worldomonation commented on packages/jest-circus-allure-reporter/src/types/interface.ts:
```diff
@@ -0,0 +1,192 @@
+/* eslint-disable require-jsdoc */
```
In the interest of getting this code reviewed sooner I have disabled the `require-jsdoc` rule for this file.

worldomonation commented on test/e2e/package.json:
```diff
@@ -14,7 +14,11 @@
 	"scripts": {
 		"encrypt-secrets": "yarn workspace @automattic/calypso-e2e run encrypt-secrets",
-		"decrypt-secrets": "yarn workspace @automattic/calypso-e2e run decrypt-secrets"
+		"decrypt-secrets": "yarn workspace @automattic/calypso-e2e run decrypt-secrets",
+		"clean": "yarn clean:packages && yarn clean:output && rm -rf ./cookies && rm -rf ./node_modules",
+		"clean:packages": "yarn workspace @automattic/jest-circus-allure-reporter clean && yarn workspace @automattic/calypso-e2e clean",
+		"clean:output": "rm -rf ./allure-results && rm -rf ./results",
+		"build": "yarn workspace @automattic/jest-circus-allure-reporter build && yarn workspace @automattic/calypso-e2e build"
```
These changes are not necessarily part of the core set of changes, but they are helpful QoL improvements.

worldomonation committed:
.teamcity/_self/lib/customBuildType/E2EBuildType.kt
- remove experimental section on adding support for allure-results output

youbrokesomething commented on packages/jest-circus-allure-reporter/README.md:
````diff
@@ -0,0 +1,152 @@
+1. **Run a Calypso end-to-end test**
+
+```shell
+yarn jest specs/onboarding/signup__free.ts
````
This spec doesn't exist anymore. This should probably be 'yarn jest specs/onboarding/ftme__write.ts'

youbrokesomething reviewed (approved):
LGTM with a minor documentation suggestion.

noahtallen commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -29,24 +34,46 @@ const supportedBrowsers = [ chromium, firefox ];
 		this.testFilePath = context.testPath;
 		this.testFilename = path.parse( context.testPath ).name;
 		this.testArtifactsPath = '';
+		this.allure = this.initializeAllureReporter( config );
```
Will this run on every e2e run? If so, do we ever want to disable it? For example, should it be disabled for local e2e? Should it only be enabled on trunk builds? etc

worldomonation replied:
That's an excellent question and one that might affect how this entire thing is structured.

The short answer is that yes, in the current implementation this will run every time E2E tests are triggered, regardless of where it's run - CI, local, wherever.

From the user's perspective, this should be transparent (since the output folder is added to `.gitignore`, and the added scripts in `package.json` will clean these out), but it will write more artifacts to the disk than the current list (browser log, video, screenshot, trace).

Conceptually this PR is for the MVP so I do plan on iterating some more on this, so perhaps not enabling Allure by default would be the right choice here. I'll also spend some time to try and make this enabled only for Pre-Release Tests.

noahtallen replied:
That sounds good to me! I imagine it could have some sort of performance impact

worldomonation replied:
I've pushed up a revised commit which should address your concerns:
- `env-variables.ts`: the `ALLURE_RESULTS_PATH` now sets to an empty string by default.
- `environment.ts`: there is a `if` clause that checks for `ALLURE_RESULTS_PATH` having a truthy value. If it does not, Allure reporter is not initialized.

I've ran this a few times locally with and without `ALLURE_RESULTS_PATH` set and have checked that everything behaves as expected.

noahtallen replied:
Nice 👍

noahtallen reviewed (commented):
I didn't really look at the jest-circus-allure-reporter code since it's basically a dependency.  I can't say I like this approach, since we now have to maintain and update this custom reporter as jest updates in the future. That could be time intensive, and I wouldn't want it to be a blocker to jest updates.

But I understand you've looked into the options and this is likely the best one :) 

Oh, one question is: where will this be deployed for use internally and what's the plan for that?

worldomonation commented:
> I can't say I like this approach, since we now have to maintain and update this custom reporter as jest updates in the future. 

Thanks for raising the concern. If I'm correct, this _should not_ affect future Jest upgrades **unless** the future Jest version introduces a new test runner, like what Jest 27 did with `jest-circus`.

Others within Automattic are using an older version of Jest (<27) to [maintain compatibility with Allure reporters](https://github.com/Automattic/wp-calypso/issues/66190#issuecomment-1211583059) but I did not want to downgrade the version of Jest we use. To develop our own reporter was the next best alternative.

> But I understand you've looked into the options and this is likely the best one :)

> Oh, one question is: where will this be deployed for use internally and what's the plan for that?

The intent for Allure reports would be to collect detailed performance data for the E2E test suites.
The resulting data can be automatically transformed into a rich dashboard like [this one](https://woocommerce.github.io/woocommerce-test-reports/daily/api/allure-report/#). The key thing here is that Allure comes with all sorts of detailed graphs and charts that we'd otherwise have to roll ourselves.

noahtallen commented:
Right -- I guess I mean, how will we go from Allure outputting data to having a website?

worldomonation commented:
> Right -- I guess I mean, how will we go from Allure outputting data to a TeamCity artifact to having a website?

Ah, sorry! My current thought is to have the TeamCity build configuration push the generated Allure report to a S3 bucket, which is what [Jetpack folks do](https://github.com/Automattic/jetpack-e2e-reports). Alternatively, it might be possible to run a GitHub action to pull the data and generate it as a static site; this is what the [WooCommerce folks do](https://github.com/woocommerce/woocommerce-test-reports).

noahtallen commented:
neat!

worldomonation committed:
packages/calypso-e2e/src/env-variables.ts
- change default value for ALLURE_RESULTS_PATH to an empty string.

packages/calypso-e2e/src/jest-playwright-config/environment.ts
- short circuit the allure reporter setup step in `initializeAllureReporter` if `ALLURE_RESULTS_PATH` is not set.
- since `this.allure` attribute can be undefined, replace calls to `this.allure` with `this.allure?`.

noahtallen reviewed (approved):
I haven't done any testing, but this seems like a fair approach overall.
